### PR TITLE
Improve Python 3 Compatibility

### PR DESCRIPTION
**Pull request by dannymilsom**
Attempt to make our codebase better support both Python 2 and 3 simultaneously.

Note this is very much a work in progress and does not promise to provide full support for Python 3 yet - but could be merged into master as we work towards this goal incrementally. For this reason I have not updated the documentation.

Summary of changes proposed in this Pull Request:
- Simultaneously support for Python 2 and 3

PR checklist:
- [ ] Updated relevant documentation
- [ ] Updated CHANGELOG.md 
- [ ] Added tests for my change

### CODE REVIEW

**Comment by davidwtbuxton** [comment]
I think all of the places where we use `dict.iteritems()` and `xrange()` should be replaced with `dict.items()` and `range()` (or for dictionaries, iterating over the dictionary directly where that makes sense). We shouldn't use the six shim for these.

`dict.iteritems()` and `xrange()` are useful for performance optimisations (memory pressure). We don't need those optimisations. Avoiding the use of six altogether would be better.

File: djangae/storage.py [hunk 1]
```diff
@@ -7,7 +7,7 @@
 try:
     from cStringIO import StringIO
 except ImportError:
-    from StringIO import StringIO
+    from io import StringIO
```
Comment (davidwtbuxton) [inline djangae/storage.py]: `StringIO` isn't used in this module. We can remove it and `re`.

**Comment by adamalton** [comment]
Just one Danny, this looks good!

I'm indifferent about the six or not-six approach to `range` and `dict.iteritems`.  I guess using less `six` would be good, but given that we're using it anyway...?  I'll let you decide!

Other than that, there's just one failing test.

**Comment by davidwtbuxton** [comment]
> I'm indifferent about the six or not-six approach to range and dict.iteritems.

Not relying on six where possible reduces complexity and will be better for the code in the long run. None of the places where the code is using `dict.iteritems()` needs to be concerned with memory use, and should have made use of `dict.items()` in the first place.

Of the code that makes use of `xrange()`, only in the deferred task stuff does it look like we may be building a sequence of potentially hundreds of items, and even in that case a list of 1,000 integers occupies less than 10 kilobytes.

Six should be used only when there isn't a more idiomatic syntax available. If it is possible to write code that works on Python 2 and 3 as-is, then we should write code that works on Python 2 and 3 as-is.

**Comment by adamalton** [comment]
Fair.

**Pull request merged**